### DISCUSSION TITLE
Web Storage API & Safari

### DISCUSSION THREAD

**Discussion by nothingrandom** [pmndrs/zustand#1912] [Bug report]
## Summary
Using the [Web Storage API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API) for persisting storage (specifically for browser extensions), @andyhails and I discovered that there is a bug in `createJSONStorage` for Safari that causes it to stringify and persist store functions (e.g. `foo()`) as objects (e.g. `foo: {}`).

This does not exist as an issue for Firefox, Chrome, or other Chromium browsers.

We found this bug to be extremely inconsistent between builds for Safari through xCode and throughout runtime, however, did find a fix through a custom `createJSONStorage`.

## Proposal
Filter functions out from being in the persisted store. Zustand handles this just fine, combining the persisted store of strings, booleans, and objects with the non-persisted store which includes those functions no longer being persisted.

`createJSONStorage.ts`
```ts
import { PersistStorage, StateStorage, StorageValue } from 'zustand/middleware';
import { StoreState } from '@/store/index';

/**
 * createJSONStorage - our own version of zustand's createLocalStorage middleware.
 *
 * There is a bug in zustand's createLocalStorage middleware for Safari that causes it to stringify and
 * persist StateStore functions (e.g. foo()) as objects (e.g. foo: {}).
 *
 * Overriding the createJSONStorage middleware with our own version that filters out functions from the state
 * mitigates this issue.
 */

/**
 * Parses a string into a StorageValue.
 * @param str
 */
const parse = (str: string | null): StorageValue<StoreState> => (str ? JSON.parse(str) : null);

/**
 * Filters out properties that cannot be persisted to storage (e.g. functions).
 * This is usually handled by the browser well but for some reason in Safari there are side effects.
 *
 * @param state
 */
const statePropertiesOnly = (state: StoreState) =>
  Object.fromEntries(
    Object.entries(state)
      .filter(([_, value]) => typeof value !== 'function'),
  );

export default (getStorage: () => StateStorage): PersistStorage<StoreState> => ({
  getItem: name => {
    const item = getStorage().getItem(name);
    return item instanceof Promise ? item.then(parse) : parse(item);
  },
  setItem: (name, newValue: StorageValue<StoreState>) =>
    getStorage().setItem(name, JSON.stringify(statePropertiesOnly(newValue.state))),
  removeItem: name => getStorage().removeItem(name),
});
```

**Comment by dai-shi** [comment]
> to stringify and persist store functions (e.g. `foo()`) as objects (e.g. `foo: {}`).

Thanks for reporting! Can you identify the problem?

This is what I tried:

## Chrome
<img width="233" alt="image" src="https://github.com/pmndrs/zustand/assets/490574/65dc0406-3451-42d6-a8f3-7d2a39f292f3">

## Safari
<img width="234" alt="image" src="https://github.com/pmndrs/zustand/assets/490574/69342ff2-6cdf-4721-824e-c3c2b5ba27d6">

**Reply by dai-shi** [reply]
Maybe we do something before `JSON.stringify`.

**Reply by nothingrandom** [reply]
Annoyingly, maybe not. We attempted testing `JSON.stringify` and `JSON.parse` between Chrome and Safari, but didn't find anything that would specifically break this in a normal page.
I _think_ it might be more related to the Web Storage API, and how Safari saves that between sessions.

If needed, I can get a mini-extension build working as a proof of the bug on this one.

**Reply by dai-shi** [reply]
So, can you reproduce the difference with the web storage api without zustand?

**Reply by andyhailscode** [reply]
Hi, 
There are a few things we've noticed (I work with @nothingrandom):
1. It's only on extension and only using `chrome.storage.local` API
2. It SEEMS OK if using session storage.
3. It's not consistent! But what seems to happen is that functions are (weirdly) converted to objects. As you demonstrated above, JSON.stringify usually strips out these functions but it seems there are times when it does not.

**Reply by dai-shi** [reply]
Thanks for the details.
So, would `partialize` option be a workaround?
https://github.com/pmndrs/zustand/blob/58947a5bf9e39aa4fc58a5356cd16eb20dbba45d/docs/integrations/persisting-store-data.md?plain=1#L80-L86

**Reply by alexarciniaga** [reply]
Is there a solution to this? I'm having the same exact issue with Safari.

**Reply by dai-shi** [reply]
I believe so, but the exact problem is still uncertain.
